File: c/src/json.c
#include "json.h"
#include "util.h"

#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    const char *text;
    size_t pos;
} Parser;

static void skip_ws(Parser *p) {
    while (p->text[p->pos]
           && (p->text[p->pos] == ' ' || p->text[p->pos] == '\n'
               || p->text[p->pos] == '\r' || p->text[p->pos] == '\t')) {
        p->pos++;
    }
}

static int peek(Parser *p, char c) {
    return p->text[p->pos] == c;
}

static int expect(Parser *p, char c) {
    skip_ws(p);
    if (!peek(p, c)) {
        return -1;
    }
    p->pos++;
    return 0;
}

static JsonValue *parse_value(Parser *p);

static char *parse_string(Parser *p) {
    if (expect(p, '"') != 0) {
        return NULL;
    }
    size_t cap = 32;
    size_t len = 0;
    char *out = malloc(cap);
    if (!out) {
        return NULL;
    }
    while (p->text[p->pos]) {
        char c = p->text[p->pos++];
        if (c == '"') {
            out[len] = '\0';
            return out;
        }
        if (c == '\\') {
            char esc = p->text[p->pos++];
            switch (esc) {
                case '"':
                case '\\':
                case '/':
                    c = esc;
                    break;
                case 'b':
                    c = '\b';
                    break;
                case 'f':
                    c = '\f';
                    break;
                case 'n':
                    c = '\n';
                    break;
                case 'r':
                    c = '\r';
                    break;
                case 't':
                    c = '\t';
                    break;
                case 'u': {
                    if (strlen(p->text + p->pos) < 4) {
                        free(out);
                        return NULL;
                    }
                    char hex[5];
                    memcpy(hex, p->text + p->pos, 4);
                    hex[4] = '\0';
                    c = (char)strtol(hex, NULL, 16);
                    p->pos += 4;
                    break;
                }
                default:
                    free(out);
                    return NULL;
            }
        }
        if (len + 1 >= cap) {
            cap *= 2;
            char *n = realloc(out, cap);
            if (!n) {
                free(out);
                return NULL;
            }
            out = n;
        }
        out[len++] = c;
    }
    free(out);
    return NULL;
}

static JsonValue *parse_number(Parser *p) {
    size_t start = p->pos;
    if (peek(p, '-')) {
        p->pos++;
    }
    while (isdigit((unsigned char)p->text[p->pos])) {
        p->pos++;
    }
    if (peek(p, '.')) {
        p->pos++;
        while (isdigit((unsigned char)p->text[p->pos])) {
            p->pos++;
        }
    }
    if (p->text[p->pos] == 'e' || p->text[p->pos] == 'E') {
        p->pos++;
        if (peek(p, '+') || peek(p, '-')) {
            p->pos++;
        }
        while (isdigit((unsigned char)p->text[p->pos])) {
            p->pos++;
        }
    }
    char *raw = wl_strndup(p->text + start, p->pos - start);
    if (!raw) {
        return NULL;
    }
    JsonValue *v = calloc(1, sizeof(JsonValue));
    if (!v) {
        free(raw);
        return NULL;
    }
    v->type = JSON_NUMBER;
    v->number_value = atof(raw);
    free(raw);
    return v;
}

static JsonValue *parse_array(Parser *p) {
    if (expect(p, '[') != 0) {
        return NULL;
    }
    JsonValue *v = calloc(1, sizeof(JsonValue));
    if (!v) {
        return NULL;
    }
    v->type = JSON_ARRAY;
    skip_ws(p);
    if (peek(p, ']')) {
        p->pos++;
        return v;
    }
    size_t cap = 0;
    while (1) {
        JsonValue *item = parse_value(p);
        if (!item) {
            json_free(v);
            return NULL;
        }
        if (v->array_len == cap) {
            size_t ncap = cap == 0 ? 4 : cap * 2;
            JsonValue **nitems = realloc(v->array_items, ncap * sizeof(JsonValue *));
            if (!nitems) {
                json_free(item);
                json_free(v);
                return NULL;
            }
            v->array_items = nitems;
            cap = ncap;
        }
        v->array_items[v->array_len++] = item;
        skip_ws(p);
        if (peek(p, ']')) {
            p->pos++;
            return v;
        }
        if (expect(p, ',') != 0) {
            json_free(v);
            return NULL;
        }
    }
}

static JsonValue *parse_object(Parser *p) {
    if (expect(p, '{') != 0) {
        return NULL;
    }
    JsonValue *v = calloc(1, sizeof(JsonValue));
    if (!v) {
        return NULL;
    }
    v->type = JSON_OBJECT;
    skip_ws(p);
    if (peek(p, '}')) {
        p->pos++;
        return v;
    }
    size_t cap = 0;
    while (1) {
        skip_ws(p);
        char *key = parse_string(p);
        if (!key) {
            json_free(v);
            return NULL;
        }
        skip_ws(p);
        if (expect(p, ':') != 0) {
            free(key);
            json_free(v);
            return NULL;
        }
        JsonValue *value = parse_value(p);
        if (!value) {
            free(key);
            json_free(v);
            return NULL;
        }
        if (v->object_len == cap) {
            size_t ncap = cap == 0 ? 4 : cap * 2;
            JsonObjectEntry *nentries = realloc(v->object_entries, ncap * sizeof(JsonObjectEntry));
            if (!nentries) {
                free(key);
                json_free(value);
                json_free(v);
                return NULL;
            }
            v->object_entries = nentries;
            cap = ncap;
        }
        v->object_entries[v->object_len].key = key;
        v->object_entries[v->object_len].value = value;
        v->object_len++;
        skip_ws(p);
        if (peek(p, '}')) {
            p->pos++;
            return v;
        }
        if (expect(p, ',') != 0) {
            json_free(v);
            return NULL;
        }
    }
}

static JsonValue *parse_value(Parser *p) {
    skip_ws(p);
    char c = p->text[p->pos];
    if (c == '{') {
        return parse_object(p);
    }
    if (c == '[') {
        return parse_array(p);
    }
    if (c == '"') {
        char *s = parse_string(p);
        if (!s) {
            return NULL;
        }
        JsonValue *v = calloc(1, sizeof(JsonValue));
        if (!v) {
            free(s);
            return NULL;
        }
        v->type = JSON_STRING;
        v->string_value = s;
        return v;
    }
    if (c == 't' && strncmp(p->text + p->pos, "true", 4) == 0) {
        p->pos += 4;
        JsonValue *v = calloc(1, sizeof(JsonValue));
        if (!v) {
            return NULL;
        }
        v->type = JSON_BOOL;
        v->bool_value = true;
        return v;
    }
    if (c == 'f' && strncmp(p->text + p->pos, "false", 5) == 0) {
        p->pos += 5;
        JsonValue *v = calloc(1, sizeof(JsonValue));
        if (!v) {
            return NULL;
        }
        v->type = JSON_BOOL;
        v->bool_value = false;
        return v;
    }
    if (c == 'n' && strncmp(p->text + p->pos, "null", 4) == 0) {
        p->pos += 4;
        JsonValue *v = calloc(1, sizeof(JsonValue));
        if (!v) {
            return NULL;
        }
        v->type = JSON_NULL;
        return v;
    }
    if (c == '-' || isdigit((unsigned char)c)) {
        return parse_number(p);
    }
    return NULL;
}

JsonValue *json_parse(const char *text) {
    Parser p = {text, 0};
    JsonValue *v = parse_value(&p);
    if (!v) {
        return NULL;
    }
    skip_ws(&p);
    if (p.text[p.pos] != '\0') {
        json_free(v);
        return NULL;
    }
    return v;
}

void json_free(JsonValue *value) {
    if (!value) {
        return;
    }
    if (value->type == JSON_STRING) {
        free(value->string_value);
    } else if (value->type == JSON_ARRAY) {
        for (size_t i = 0; i < value->array_len; i++) {
            json_free(value->array_items[i]);
        }
        free(value->array_items);
    } else if (value->type == JSON_OBJECT) {
        for (size_t i = 0; i < value->object_len; i++) {
            free(value->object_entries[i].key);
            json_free(value->object_entries[i].value);
        }
        free(value->object_entries);
    }
    free(value);
}

const JsonValue *json_object_get(const JsonValue *obj, const char *key) {
    if (!obj || obj->type != JSON_OBJECT) {
        return NULL;
    }
    for (size_t i = 0; i < obj->object_len; i++) {
        if (strcmp(obj->object_entries[i].key, key) == 0) {
            return obj->object_entries[i].value;
        }
    }
    return NULL;
}

File: fortran/src/json.f90
! Minimal JSON parser for --template (ISO Fortran only).
module wl_json
  use wl_util
  implicit none
  private

  public :: JSON_NULL, JSON_BOOL, JSON_NUMBER, JSON_STRING, JSON_ARRAY, JSON_OBJECT
  public :: json_value, json_object_entry
  public :: json_parse, json_free, json_object_get

  integer, parameter :: JSON_NULL = 0
  integer, parameter :: JSON_BOOL = 1
  integer, parameter :: JSON_NUMBER = 2
  integer, parameter :: JSON_STRING = 3
  integer, parameter :: JSON_ARRAY = 4
  integer, parameter :: JSON_OBJECT = 5

  type :: json_value
    integer :: value_type = JSON_NULL
    logical :: bool_value = .false.
    real(kind=kind(1.0d0)) :: number_value = 0.0d0
    character(len=:), allocatable :: string_value
    type(json_value), pointer :: array_items(:) => null()
    integer :: array_len = 0
    type(json_object_entry), pointer :: object_entries(:) => null()
    integer :: object_len = 0
  end type json_value

  type :: json_object_entry
    character(len=:), allocatable :: key
    type(json_value), pointer :: value => null()
  end type json_object_entry

  type :: parser_t
    character(len=:), allocatable :: text
    integer :: pos = 1
  end type parser_t

contains

  recursive subroutine json_free_inplace(value)
    type(json_value), intent(inout) :: value
    integer :: i
    if (value%value_type == JSON_STRING) then
      if (allocated(value%string_value)) deallocate(value%string_value)
    else if (value%value_type == JSON_ARRAY) then
      if (associated(value%array_items)) then
        do i = 1, value%array_len
          call json_free_inplace(value%array_items(i))
        end do
        deallocate(value%array_items)
      end if
    else if (value%value_type == JSON_OBJECT) then
      if (associated(value%object_entries)) then
        do i = 1, value%object_len
          if (allocated(value%object_entries(i)%key)) &
            deallocate(value%object_entries(i)%key)
          if (associated(value%object_entries(i)%value)) then
            call json_free(value%object_entries(i)%value)
          end if
        end do
        deallocate(value%object_entries)
      end if
    end if
    value%value_type = JSON_NULL
    value%array_len = 0
    value%object_len = 0
    value%array_items => null()
    value%object_entries => null()
  end subroutine json_free_inplace

  recursive subroutine json_free(value)
    type(json_value), pointer, intent(inout) :: value
    if (.not. associated(value)) return
    call json_free_inplace(value)
    deallocate(value)
    value => null()
  end subroutine json_free

  function json_object_get(obj, key) result(val)
    type(json_value), intent(in) :: obj
    character(len=*), intent(in) :: key
    type(json_value), pointer :: val
    integer :: i
    val => null()
    if (obj%value_type /= JSON_OBJECT) return
    if (.not. associated(obj%object_entries)) return
    do i = 1, obj%object_len
      if (obj%object_entries(i)%key == key) then
        val => obj%object_entries(i)%value
        return
      end if
    end do
  end function json_object_get

  subroutine skip_ws(p)
    type(parser_t), intent(inout) :: p
    do while (p%pos <= len(p%text))
      if (.not. is_space(p%text(p%pos:p%pos))) exit
      p%pos = p%pos + 1
    end do
  end subroutine skip_ws

  logical function peek(p, c) result(ok)
    type(parser_t), intent(in) :: p
    character(len=1), intent(in) :: c
    ok = (p%pos <= len(p%text) .and. p%text(p%pos:p%pos) == c)
  end function peek

  logical function expect(p, c) result(ok)
    type(parser_t), intent(inout) :: p
    character(len=1), intent(in) :: c
    call skip_ws(p)
    ok = peek(p, c)
    if (ok) p%pos = p%pos + 1
  end function expect

  recursive function parse_value(p, ok) result(val)
    type(parser_t), intent(inout) :: p
    logical, intent(out) :: ok
    type(json_value), pointer :: val
    character(len=1) :: c
    character(len=:), allocatable :: s
    ok = .false.
    val => null()
    call skip_ws(p)
    if (p%pos > len(p%text)) return
    c = p%text(p%pos:p%pos)
    if (c == '{') then
      val => parse_object(p, ok)
      return
    end if
    if (c == '[') then
      val => parse_array(p, ok)
      return
    end if
    if (c == '"') then
      s = parse_string(p, ok)
      if (.not. ok) return
      allocate(val)
      val%value_type = JSON_STRING
      val%string_value = s
      return
    end if
    if (c == 't' .and. p%pos + 3 <= len(p%text) .and. p%text(p%pos:p%pos+3) == 'true') then
      p%pos = p%pos + 4
      allocate(val)
      val%value_type = JSON_BOOL
      val%bool_value = .true.
      ok = .true.
      return
    end if
    if (c == 'f' .and. p%pos + 4 <= len(p%text) .and. p%text(p%pos:p%pos+4) == 'false') then
      p%pos = p%pos + 5
      allocate(val)
      val%value_type = JSON_BOOL
      val%bool_value = .false.
      ok = .true.
      return
    end if
    if (c == 'n' .and. p%pos + 3 <= len(p%text) .and. p%text(p%pos:p%pos+3) == 'null') then
      p%pos = p%pos + 4
      allocate(val)
      val%value_type = JSON_NULL
      ok = .true.
      return
    end if
    if (c == '-' .or. is_digit(c)) then
      val => parse_number(p, ok)
      return
    end if
  end function parse_value

  function parse_string(p, ok) result(out)
    type(parser_t), intent(inout) :: p
    logical, intent(out) :: ok
    character(len=:), allocatable :: out
    character(len=1) :: c, esc
    character(len=4) :: hex
    integer :: code, ios
    ok = .false.
    out = ''
    if (.not. expect(p, '"')) return
    do while (p%pos <= len(p%text))
      c = p%text(p%pos:p%pos)
      p%pos = p%pos + 1
      if (c == '"') then
        ok = .true.
        return
      end if
      if (c == '\') then
        if (p%pos > len(p%text)) return
        esc = p%text(p%pos:p%pos)
        p%pos = p%pos + 1
        select case (esc)
        case ('"', '\', '/')
          out = out // esc
        case ('b')
          out = out // achar(8)
        case ('f')
          out = out // achar(12)
        case ('n')
          out = out // achar(10)
        case ('r')
          out = out // achar(13)
        case ('t')
          out = out // achar(9)
        case ('u')
          if (p%pos + 3 > len(p%text)) return
          hex = p%text(p%pos:p%pos+3)
          p%pos = p%pos + 4
          read(hex, '(Z4)', iostat=ios) code
          if (ios /= 0) return
          out = out // achar(iand(code, 255))
        case default
          return
        end select
      else
        out = out // c
      end if
    end do
  end function parse_string

  function parse_number(p, ok) result(val)
    type(parser_t), intent(inout) :: p
    logical, intent(out) :: ok
    type(json_value), pointer :: val
    integer :: start_pos, ios
    character(len=:), allocatable :: raw
    real(kind=kind(1.0d0)) :: num
    ok = .false.
    val => null()
    start_pos = p%pos
    if (peek(p, '-')) p%pos = p%pos + 1
    do while (p%pos <= len(p%text) .and. is_digit(p%text(p%pos:p%pos)))
      p%pos = p%pos + 1
    end do
    if (peek(p, '.')) then
      p%pos = p%pos + 1
      do while (p%pos <= len(p%text) .and. is_digit(p%text(p%pos:p%pos)))
        p%pos = p%pos + 1
      end do
    end if
    if (p%pos <= len(p%text) .and. &
        (p%text(p%pos:p%pos) == 'e' .or. p%text(p%pos:p%pos) == 'E')) then
      p%pos = p%pos + 1
      if (peek(p, '+') .or. peek(p, '-')) p%pos = p%pos + 1
      do while (p%pos <= len(p%text) .and. is_digit(p%text(p%pos:p%pos)))
        p%pos = p%pos + 1
      end do
    end if
    raw = p%text(start_pos:p%pos-1)
    read(raw, *, iostat=ios) num
    if (ios /= 0) return
    allocate(val)
    val%value_type = JSON_NUMBER
    val%number_value = num
    ok = .true.
  end function parse_number

  recursive function parse_array(p, ok) result(val)
    type(parser_t), intent(inout) :: p
    logical, intent(out) :: ok
    type(json_value), pointer :: val
    type(json_value), pointer :: item
    type(json_value), pointer :: tmp(:)
    integer :: cap, i
    ok = .false.
    val => null()
    if (.not. expect(p, '[')) return
    allocate(val)
    val%value_type = JSON_ARRAY
    val%array_len = 0
    call skip_ws(p)
    if (peek(p, ']')) then
      p%pos = p%pos + 1
      ok = .true.
      return
    end if
    cap = 0
    do
      item => parse_value(p, ok)
      if (.not. ok .or. .not. associated(item)) then
        call json_free(val)
        return
      end if
      if (val%array_len >= cap) then
        if (cap == 0) then
          cap = 4
        else
          cap = cap * 2
        end if
        allocate(tmp(cap))
        do i = 1, val%array_len
          tmp(i) = val%array_items(i)
        end do
        if (associated(val%array_items)) deallocate(val%array_items)
        val%array_items => tmp
      end if
      val%array_len = val%array_len + 1
      val%array_items(val%array_len) = item
      ! item was a heap root; components moved via assignment — free shell
      deallocate(item)
      call skip_ws(p)
      if (peek(p, ']')) then
        p%pos = p%pos + 1
        ok = .true.
        return
      end if
      if (.not. expect(p, ',')) then
        call json_free(val)
        ok = .false.
        return
      end if
    end do
  end function parse_array

  recursive function parse_object(p, ok) result(val)
    type(parser_t), intent(inout) :: p
    logical, intent(out) :: ok
    type(json_value), pointer :: val
    type(json_value), pointer :: value
    type(json_object_entry), pointer :: tmp(:)
    character(len=:), allocatable :: key
    integer :: cap, i
    ok = .false.
    val => null()
    if (.not. expect(p, '{')) return
    allocate(val)
    val%value_type = JSON_OBJECT
    val%object_len = 0
    call skip_ws(p)
    if (peek(p, '}')) then
      p%pos = p%pos + 1
      ok = .true.
      return
    end if
    cap = 0
    do
      call skip_ws(p)
      key = parse_string(p, ok)
      if (.not. ok) then
        call json_free(val)
        return
      end if
      call skip_ws(p)
      if (.not. expect(p, ':')) then
        call json_free(val)
        ok = .false.
        return
      end if
      value => parse_value(p, ok)
      if (.not. ok .or. .not. associated(value)) then
        call json_free(val)
        ok = .false.
        return
      end if
      if (val%object_len >= cap) then
        if (cap == 0) then
          cap = 4
        else
          cap = cap * 2
        end if
        allocate(tmp(cap))
        do i = 1, val%object_len
          tmp(i)%key = val%object_entries(i)%key
          tmp(i)%value => val%object_entries(i)%value
        end do
        if (associated(val%object_entries)) deallocate(val%object_entries)
        val%object_entries => tmp
      end if
      val%object_len = val%object_len + 1
      val%object_entries(val%object_len)%key = key
      val%object_entries(val%object_len)%value => value
      call skip_ws(p)
      if (peek(p, '}')) then
        p%pos = p%pos + 1
        ok = .true.
        return
      end if
      if (.not. expect(p, ',')) then
        call json_free(val)
        ok = .false.
        return
      end if
    end do
  end function parse_object

  function json_parse(text) result(val)
    character(len=*), intent(in) :: text
    type(json_value), pointer :: val
    type(parser_t) :: p
    logical :: ok
    p%text = text
    p%pos = 1
    val => parse_value(p, ok)
    if (.not. ok .or. .not. associated(val)) then
      val => null()
      return
    end if
    call skip_ws(p)
    if (p%pos <= len(p%text)) then
      call json_free(val)
      return
    end if
  end function json_parse

end module wl_json
